src/System.Buffers/src/System/Buffers/DefaultArrayPoolBucket.cs

```diff
@@ -7,114 +7,108 @@
-            _exhaustedEventSent = false;
-            _poolId = poolId;
-        }
+            internal readonly int _bufferLength;
+            private readonly T[][] _buffers;
+            private readonly int _poolId;
 
-        /// <summary>
-        /// Returns an array from the Bucket sized according to the Bucket size.
-        /// If the Bucket is empty, null is returned.
-        /// </summary>
-        /// <returns>Returns a valid buffer when the bucket has free buffers; otherwise, returns null</returns>
-        internal T[] Rent()
-        {
-            T[] buffer = null;
+            private SpinLock _lock; // do not make this readonly; it's a mutable struct
+            private int _index;
+
+            /// <summary>
+            /// Creates the pool with numberOfBuffers arrays where each buffer is of bufferLength length.
+            /// </summary>
+            internal Bucket(int bufferLength, int numberOfBuffers, int poolId)
+            {
+                _lock = new SpinLock(Debugger.IsAttached); // only enable thread tracking if debugger is attached; it adds non-trivial overheads to Enter/Exit
+                _buffers = new T[numberOfBuffers][];
+                _bufferLength = bufferLength;
+                _poolId = poolId;
+            }
+
+            /// <summary>Gets an ID for the bucket to use with events.</summary>
+            internal int Id => GetHashCode();
 
-            // Use a SpinLock since it is super lightweight
-            // and our lock is very short lived. Wrap in try-finally
-            // to protect against thread-aborts
-            bool taken = false;
-            bool alloc = false;
-            try
+            /// <summary>Takes an array from the bucket.  If the bucket is empty, returns null.</summary>
+            internal T[] Rent()
             {
-                _lock.Enter(ref taken);
+                T[][] buffers = _buffers;
+                T[] buffer = null;
 
-                // Check if all of our buffers have been rented
-                if (_index < _data.Length)
+                // While holding the lock, grab whatever is at the next available index and
+                // update the index.  We do as little work as possible while holding the spin
+                // lock to minimize contention with other threads.  The try/finally is
+                // necessary to properly handle thread aborts on platforms which have them.
+                bool lockTaken = false, allocateBuffer = false;
+                try
                 {
-                    buffer = _data[_index];
-                    alloc = buffer == null;
-                    _data[_index++] = null;
+                    _lock.Enter(ref lockTaken);
+
+                    if (_index < buffers.Length)
+                    {
+                        buffer = buffers[_index];
+                        buffers[_index++] = null;
+                        allocateBuffer = buffer == null;
```

Comment:
What is the purpose of the allocateBuffer bool? Couldn't we just check buffer == null on line 66?